### PR TITLE
Support beacon chain

### PR DESCRIPTION
**Pull request by KirillPamPam**
Now we can support chains with rest API. And the first supported chain is beacon chain.
This version is basic version and it includes:
- parsing yaml and creating a rest chain with specific Reader where we parse a rest request and then stream a response as is
- validations of beacon chain - sync, health, peer count
- poll latest head and its parsing
- detect client type version as label
- all rest methods

Also I renamed all objects and classes with `JsonRpc` prefix because we need something more abstract if we have another type of requests or something else.

### CODE REVIEW

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/CallParams.kt [hunk 1]
```diff
@@ -1,11 +1,76 @@
...
+) : CallParams {
+    companion object {
+        fun emptyParams() = RestParams(emptyMap(), emptyMap(), emptyList(), ByteArray(0))
+    }
+
+    override fun toJson(id: Int, method: String): ByteArray {
+        return payload
+    }
+
+    // we need equals and hashCode because of having ByteArray
```
Comment (a10zn8) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/CallParams.kt:56]: Are we using comparison of RestParams?

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/beaconchain/BeaconChainSpecific.kt [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            RestParams.emptyParams(),
+        )
+    }
+
+    override fun parseBlock(data: ByteArray, upstreamId: String): BlockContainer {
+        val block = Global.objectMapper.readValue(data, BeaconChainBlock::class.java).data.message
+
+        return BlockContainer(
+            height = block.slot.toLong(),
+            hash = BlockId.fromBase64(block.body.executionPayload.hash),
```
[the remark this answers is not in the document]
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/beaconchain/BeaconChainSpecific.kt:36]: Let me doublecheck

File: buildSrc/src/main/kotlin/chainsconfig.codegen.gradle.kts [hunk 3]
```diff
@@ -119,9 +133,18 @@ open class CodeGen(private val config: ChainsConfig) {
             "polkadot" -> "BlockchainType.POLKADOT"
             "solana" -> "BlockchainType.SOLANA"
             "near" -> "BlockchainType.NEAR"
+            "beacon-chain" -> "BlockchainType.BEACON_CHAIN"
             else -> throw IllegalArgumentException("unknown blockchain type $type")
         }
     }
+
+    private fun apiType(apiType: String?): String {
```
[the remark this answers is not in the document]
Comment (a10zn8) [reply buildSrc/src/main/kotlin/chainsconfig.codegen.gradle.kts:141]: So, if we don't have variation here at all (we cannot use ethereum + rest_api for example) I suggest store API type in Blockchain type like:
```kotlin
enum class BlockchainType(val apiType: ApiType) {
    UNKNOWN(JSONRPC),
    BITCOIN(JSONRPC), 
    ETHEREUM(JSONRPC),
    STARKNET(JSONRPC), 
    POLKADOT(JSONRPC), 
    SOLANA(JSONRPC), 
    NEAR(JSONRPC),
    ETHEREUM_BEACON(REST);
}

enum class ApiType {
    JSONRPC, REST
}

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/beaconchain/BeaconChainSpecific.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/beaconchain/BeaconChainSpecific.kt]: Fixed - combine 2 methods to get the full data of a beacon block

File: foundation/src/main/resources/chains.yaml [hunk 4]
```diff
@@ -1361,3 +1361,36 @@ chain-settings:
           grpcId: 1052
           short-names: [dymension]
           chain-id: 0x44c
+    - id: beacon-chain
```
Comment (a10zn8) [inline foundation/src/main/resources/chains.yaml:1364]: maybe ethereum-beacon-chain too?

File: foundation/src/main/resources/chains.yaml [hunk 5]
```diff
@@ -1361,3 +1361,36 @@ chain-settings:
...
+          priority: 1
+          short-names: [beacon-chain-goerli]
+        - id: Sepolia
+          code: SEPOLIA_BEACON_CHAIN
+          grpcId: 10070
+          priority: 10
+          chain-id: 0x0
+          short-names: [beacon-chain-sepolia]
+        - id: Holesky
+          code: ETHEREUM_HOLESKY_BEACON_CHAIN
```
Comment (a10zn8) [inline foundation/src/main/resources/chains.yaml:1392]: ETHEREUM_HOLESKY_BEACON_CHAIN but SEPOLIA_BEACON_CHAIN - I think we should choose one style

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/beaconchain/BeaconChainSpecific.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (a10zn8) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/beaconchain/BeaconChainSpecific.kt]: found in api doc - 

> string(path) | Block identifier. Can be one of: "head" (canonical head in node's view), "genesis", "finalized", <slot>, <hex encoded blockRoot with 0x prefix>.

So, if block == blockRoot, then it looks like [/eth/v1/beacon/headers/{block_id}](https://ethereum.github.io/beacon-APIs/#/Beacon/getBlockHeader) returns all data in single request, or am I wrong?

File: foundation/src/main/resources/chains.yaml (hunk 4 above)
Comment (KirillPamPam) [reply foundation/src/main/resources/chains.yaml]: Renamed to `eth-beacon-chain`

File: foundation/src/main/resources/chains.yaml (hunk 5 above)
Comment (KirillPamPam) [reply foundation/src/main/resources/chains.yaml]: Renamed to `HOLESKY_BEACON_CHAIN`

**Review by a10zn8 — APPROVED** [review]

**Pull request merged**